modules/fileout_netcdf/FONcArrayStructure.cc

```diff
@@ -0,0 +1,220 @@
+// FONcArrayStructure.cc
+
+// This file is part of BES Netcdf File Out Module
+
+// // Copyright (c) The HDF Group, Inc. and OPeNDAP, Inc.
+//
+// This is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License as published by the Free
+// Software Foundation; either version 2.1 of the License, or (at your
+// option) any later version.
+//
+// This software is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+// License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+//
+// You can contact OPeNDAP, Inc. at PO Box 112, Saunderstown, RI. 02874-0112.
+// You can contact The HDF Group, Inc. at 410 E University Ave,
+// Suite 200, Champaign, IL 61820
+
+#include <BESInternalError.h>
+#include <BESDebug.h>
+
+#include "FONcArrayStructure.h"
+#include "FONcUtils.h"
+#include "FONcAttributes.h"
+
+/** @brief Constructor for FONcArrayStructure that takes a DAP Array Structure
+ *
+ * This constructor takes a DAP BaseType and makes sure that it is a DAP Array
+ * Structure instance. If not, it throws an exception
+ *
+ * @param b A DAP BaseType that should be an array of Structure
+ * @throws BESInternalError if the BaseType is not a an array
+ */
+FONcArrayStructure::FONcArrayStructure(BaseType *b) :
+    FONcBaseType()
+{
+    _as = dynamic_cast<Array *>(b);
+    if (!_as) {
+        auto s = (string) "File out netcdf, array of structure was passed a " + "variable that is not an array ";
+        throw BESInternalError(s, __FILE__, __LINE__);
+    }
+    if (_as->var()->type() != dods_structure_c) {
+        string s = (string) "File out netcdf, array of structure was passed a " + "variable that is not a structure ";
+        throw BESInternalError(s, __FILE__, __LINE__);
+    }
+}
+
+/** @brief Destructor that cleans up the array of structure
+ *
+ * Delete each of the FONcBaseType instances that is a part of this
+ * structure.
+ */
+FONcArrayStructure::~FONcArrayStructure()
+{
+
+    // Adapt the code from FONcStructure.cc
+    bool done = false;
+    while (!done) {
```

Comment:
I think this while loop can be replaced by

for (auto &var: _vars) {
    delete var;
}

I don't thiink the code needs to call erase() because this is the destructor for the FONcArrayStructure object so its fields won't exist when this returns.

Reply:
Your way is much elegant. I simply copied the destructor from FONcStructure.cc. I change the code to use your way.